File: LeetCode/Problems/3. Longest Substring Without Repeating Characters.py
#First try
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start_pos = 0
        end_pos = 0
        max_count = 0

        
        for idx, ch in enumerate(s):
            if ch not in s[start_pos:idx]:
                end_pos = idx
            
            elif ch in s[start_pos:idx]:
                while(ch in s[start_pos:idx]):
                    start_pos +=1
                end_pos = idx
            
            if(end_pos - start_pos + 1 > max_count):
                max_count = end_pos - start_pos + 1
                
        
        return max_count

#hash map
#Time: O(n)
#Space: O(n)
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        
        lookUp = {}
        longest = 0
        start = 0
        for i,c in enumerate(s):
            if c in lookUp:
                #we only look at the current window, it could be smaller than start
                start = max(start,lookUp[c]+1)
                
            longest = max(longest, i - start+1)  
            #set to the latest
            lookUp[c] = i
        

        return longest

File: LeetCode/Problems/C++/3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(!s.size()) return 0;
        
        unordered_map<char,int> lookup;
        int max_length = INT_MIN, start = 0;
        for(int i=0;i<s.size();i++){
            if(lookup.find(s[i])!=lookup.end()){
                start = max(start,lookup[s[i]]+1);
            }
            max_length = max(max_length, i-start+1);
            lookup[s[i]] = i;
        }
        return max_length;
    }
};